### PR TITLE
prevent NPE when trying to uncompress a null BytesReference

### PR DESCRIPTION
**Pull request by jaymode**
This commit checks for a null BytesReference being passed to CompressorFactory#uncompressIfNeeded and simply returns null. Previously this would have resulted in a NPE. While this does seem internal at first glance, it can affect user code as a GetResponse could trigger this when the document is missing.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/common/compress/CompressorFactory.java [hunk 1]
```diff
@@ -70,6 +70,10 @@ private static boolean isAncient(BytesReference bytes) {
      * Uncompress the provided data, data can be detected as compressed using {@link #isCompressed(BytesReference)}.
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/compress/CompressorFactory.java]: I wonder if it makes more sense to do this in `compressor`?

**Review by nik9000 — APPROVED** [review]
I'm fine with the change so long as we add something to the javadoc. I don't know that it is a bug, but it is reasonable enough for it to handle null.

**Comment by javanna** [comment]
I wonder what triggered this. Ideally one shouldn't try to retrieve the document if the `GetResponse#isExists` returns false? NPE is always bad, but I wonder if we should add this leniency to the CompressorFactory. I'd rather have there a check that the argument is not null, not sure if this can only be a dream or also reality though :)

**Comment by jaymode** [comment]
> I wonder what triggered this.

`GetResponse#isExists` was checked but a bug caused this to be executed. Given that this is in our current user code (transport client), I really think we should not have a NPE here especially since our javadocs don't state anything about checking `isExists` before calling `getSourceAsBytesRef`.

>  I'd rather have there a check that the argument is not null

I can move this up to guard against this in `GetResponse` only and require a non-null argument in CompressorFactory. @javanna wdyt?

**Comment by jaymode** [comment]
> I am not sure what can of worms this last bit opens up, but fixing that would be fantastic :)

Surprisingly this has not opened a can of worms. I moved the null check from compressor factory to `GetResult` as this is more in line with the other methods in `GetResult`

**Review by javanna — APPROVED** [review]
LGTM thanks a lot @jaymode

**Pull request merged**